File: src/amedas_rainfall/statistics/annual_maxima.py
"""3種類の年区切りによる年最大値・データ完全性の計算（10節）。"""

from __future__ import annotations

import datetime as dt

import numpy as np
import pandas as pd

from amedas_rainfall.models import AnnualCompleteness, YearBoundaryDefinition

CALENDAR_YEAR = YearBoundaryDefinition(key="calendar", label="暦年", start_month=1, start_day=1)
FISCAL_YEAR = YearBoundaryDefinition(key="fiscal", label="年度", start_month=4, start_day=1)
JUNE_START_YEAR = YearBoundaryDefinition(
    key="june_start", label="6月始まり年", start_month=6, start_day=1
)

ALL_YEAR_BOUNDARIES: dict[str, YearBoundaryDefinition] = {
    "calendar": CALENDAR_YEAR,
    "fiscal": FISCAL_YEAR,
    "june_start": JUNE_START_YEAR,
}


def _start_year_for_timestamp(ts: pd.Timestamp, boundary: YearBoundaryDefinition) -> int:
    if (ts.month, ts.day) >= (boundary.start_month, boundary.start_day):
        return ts.year
    return ts.year - 1


def year_label(start_year: int, boundary: YearBoundaryDefinition) -> str:
    if boundary.key == "calendar":
        return f"{start_year}年"
    if boundary.key == "fiscal":
        return f"{start_year}年度"
    if boundary.key == "june_start":
        return f"{start_year}年6月始まり"
    return f"{start_year}年（{boundary.label}）"


def year_window(start_year: int, boundary: YearBoundaryDefinition, tz: str = "Asia/Tokyo") -> tuple[pd.Timestamp, pd.Timestamp]:
    """年区分の[開始, 終了)を返す（終了は排他的な次年区分の開始時刻）。"""
    start = pd.Timestamp(
        year=start_year, month=boundary.start_month, day=boundary.start_day, hour=0, tz=tz
    )
    end = pd.Timestamp(
        year=start_year + 1, month=boundary.start_month, day=boundary.start_day, hour=0, tz=tz
    )
    return start, end


def assign_year_labels(index: pd.DatetimeIndex, boundary: YearBoundaryDefinition) -> pd.Series:
    """各時刻がどの年区分に属するかのラベル系列を返す。"""
    start_years = [_start_year_for_timestamp(ts, boundary) for ts in index]
    labels = [year_label(y, boundary) for y in start_years]
    return pd.Series(labels, index=index, name="year_label")


def calculate_annual_maxima(
    series: pd.Series,
    boundary: YearBoundaryDefinition,
) -> pd.DataFrame:
    """指定した年区分での年最大値とその発生日時を計算する。"""
    if series.empty:
        return pd.DataFrame(columns=["year_label", "start_year", "max_value", "max_datetime"])

    labels = assign_year_labels(series.index, boundary)
    df = pd.DataFrame({"value": series, "year_label": labels})

    records = []
    for lbl, group in df.groupby("year_label", sort=False):
        valid = group["value"].dropna()
        if valid.empty:
            max_value = np.nan
            max_dt: pd.Timestamp | None = None
        else:
            max_dt = valid.idxmax()
            max_value = valid.loc[max_dt]
        start_year = int(str(lbl).split("年")[0])
        records.append(
            {
                "year_label": lbl,
                "start_year": start_year,
                "max_value": max_value,
                "max_datetime": max_dt,
            }
        )
    result = pd.DataFrame(records).sort_values("start_year").reset_index(drop=True)
    return result


def calculate_annual_completeness(
    valid_mask: pd.Series,
    boundary: YearBoundaryDefinition,
    state_reset_mask: pd.Series | None = None,
    completeness_threshold_percent: float = 95.0,
    data_start: pd.Timestamp | None = None,
    data_end: pd.Timestamp | None = None,
    now: pd.Timestamp | None = None,
) -> list[AnnualCompleteness]:
    """年区分ごとのデータ完全性を評価し、既定の採否判定を付ける。

    Args:
        valid_mask: 有効（欠測でない）時刻はTrueの真偽値系列（連続1時間インデックス）。
        boundary: 年区切り定義。
        state_reset_mask: 状態量再初期化（欠測明け）が発生した時刻でTrueの系列。
        completeness_threshold_percent: 採用可否の既定閾値。
        data_start: 観測データの実際の開始時刻（先頭不完全年判定に使用）。
        data_end: 観測データの実際の終了時刻。
        now: 現在時刻（最新未終了年区分の判定に使用）。
    """
    if valid_mask.empty:
        return []

    index = valid_mask.index
    tz = index.tz
    data_start = data_start or index.min()
    data_end = data_end or index.max()
    now = now or pd.Timestamp.now(tz=tz)

    start_year_min = _start_year_for_timestamp(data_start, boundary)
    start_year_max = _start_year_for_timestamp(data_end, boundary) + 1

    results: list[AnnualCompleteness] = []
    for start_year in range(start_year_min, start_year_max + 1):
        win_start, win_end = year_window(start_year, boundary, tz=str(tz))
        mask = (index >= win_start) & (index < win_end)
        expected_hours = int(mask.sum())
        if expected_hours == 0:
            continue
        valid_hours = int(valid_mask.loc[mask].sum())
        missing_hours = expected_hours - valid_hours
        completeness = 100.0 * valid_hours / expected_hours if expected_hours else 0.0
        has_reset = bool(state_reset_mask.loc[mask].any()) if state_reset_mask is not None else False

        reasons: list[str] = []
        is_incomplete_start = win_start < data_start <= win_end
        is_ongoing_latest = win_end > now
        if is_incomplete_start:
            reasons.append("観測開始を含む不完全年")
        if is_ongoing_latest:
            reasons.append("実行時点で終了していない最新年区分")
        if completeness < completeness_threshold_percent:
            reasons.append(f"データ完全率が{completeness_threshold_percent}%未満")
        if has_reset:
            reasons.append("大きな欠測により状態量が再初期化された区間を含む")

        is_eligible = len(reasons) == 0

        results.append(
            AnnualCompleteness(
                year_label=year_label(start_year, boundary),
                start_datetime=win_start.to_pydatetime(),
                end_datetime=win_end.to_pydatetime(),
                expected_hours=expected_hours,
                valid_hours=valid_hours,
                missing_hours=missing_hours,
                completeness_percent=completeness,
                has_state_reset=has_reset,
                is_eligible_default=is_eligible,
                exclusion_reasons=reasons,
            )
        )
    return results

File: core/src/annual_maxima.rs
//! 3種類の年区切りによる年最大値・データ完全性の計算（docs/calculation_method.md 8節）。
//!
//! Asia/Tokyoはサマータイムを持たないため、固定UTC+9オフセットとして扱う。

use chrono::{DateTime, FixedOffset, TimeZone};

pub struct YearBoundary {
    pub key: &'static str,
    pub start_month: u32,
    pub start_day: u32,
}

pub const CALENDAR_YEAR: YearBoundary = YearBoundary { key: "calendar", start_month: 1, start_day: 1 };
pub const FISCAL_YEAR: YearBoundary = YearBoundary { key: "fiscal", start_month: 4, start_day: 1 };
pub const JUNE_START_YEAR: YearBoundary = YearBoundary { key: "june_start", start_month: 6, start_day: 1 };

pub fn jst_offset() -> FixedOffset {
    FixedOffset::east_opt(9 * 3600).unwrap()
}

pub fn year_window(start_year: i32, boundary: &YearBoundary) -> (DateTime<FixedOffset>, DateTime<FixedOffset>) {
    let tz = jst_offset();
    let start = tz
        .with_ymd_and_hms(start_year, boundary.start_month, boundary.start_day, 0, 0, 0)
        .unwrap();
    let end = tz
        .with_ymd_and_hms(start_year + 1, boundary.start_month, boundary.start_day, 0, 0, 0)
        .unwrap();
    (start, end)
}

pub fn year_label(start_year: i32, boundary: &YearBoundary) -> String {
    match boundary.key {
        "calendar" => format!("{start_year}年"),
        "fiscal" => format!("{start_year}年度"),
        "june_start" => format!("{start_year}年6月始まり"),
        _ => format!("{start_year}年"),
    }
}

fn start_year_for_timestamp(ts: &DateTime<FixedOffset>, boundary: &YearBoundary) -> i32 {
    use chrono::Datelike;
    let md = (ts.month(), ts.day());
    if md >= (boundary.start_month, boundary.start_day) {
        ts.year()
    } else {
        ts.year() - 1
    }
}

pub struct AnnualMaximum {
    pub year_label: String,
    pub start_year: i32,
    pub max_value: f64,
    pub max_datetime: Option<DateTime<FixedOffset>>,
}

/// 指定した年区分での年最大値とその発生日時を計算する。
/// 同値の場合は最初に出現した時刻を採用する（pandas `idxmax` と同じ挙動）。
pub fn calculate_annual_maxima(
    index: &[DateTime<FixedOffset>],
    values: &[f64],
    boundary: &YearBoundary,
) -> Vec<AnnualMaximum> {
    use std::collections::BTreeMap;
    let mut groups: BTreeMap<i32, Vec<(DateTime<FixedOffset>, f64)>> = BTreeMap::new();
    for (ts, &v) in index.iter().zip(values.iter()) {
        let sy = start_year_for_timestamp(ts, boundary);
        groups.entry(sy).or_default().push((*ts, v));
    }
    groups
        .into_iter()
        .map(|(sy, entries)| {
            let mut max_v = f64::NAN;
            let mut max_dt: Option<DateTime<FixedOffset>> = None;
            for (ts, v) in entries {
                if v.is_nan() {
                    continue;
                }
                if max_dt.is_none() || v > max_v {
                    max_v = v;
                    max_dt = Some(ts);
                }
            }
            AnnualMaximum {
                year_label: year_label(sy, boundary),
                start_year: sy,
                max_value: max_v,
                max_datetime: max_dt,
            }
        })
        .collect()
}

pub struct AnnualCompleteness {
    pub year_label: String,
    pub expected_hours: i64,
    pub valid_hours: i64,
    pub missing_hours: i64,
    pub completeness_percent: f64,
    pub has_state_reset: bool,
    pub is_eligible_default: bool,
    pub exclusion_reasons: Vec<String>,
}

/// 年区分ごとのデータ完全性を評価し、既定の採否判定を付ける。
#[allow(clippy::too_many_arguments)]
pub fn calculate_annual_completeness(
    index: &[DateTime<FixedOffset>],
    valid_mask: &[bool],
    boundary: &YearBoundary,
    state_reset_mask: Option<&[bool]>,
    completeness_threshold_percent: f64,
    data_start: DateTime<FixedOffset>,
    data_end: DateTime<FixedOffset>,
    now: DateTime<FixedOffset>,
) -> Vec<AnnualCompleteness> {
    if index.is_empty() {
        return vec![];
    }
    let start_year_min = start_year_for_timestamp(&data_start, boundary);
    let start_year_max = start_year_for_timestamp(&data_end, boundary) + 1;

    let mut results = Vec::new();
    for start_year in start_year_min..=start_year_max {
        let (win_start, win_end) = year_window(start_year, boundary);
        let mut expected_hours: i64 = 0;
        let mut valid_hours: i64 = 0;
        let mut has_reset = false;
        for (i, ts) in index.iter().enumerate() {
            if *ts >= win_start && *ts < win_end {
                expected_hours += 1;
                if valid_mask[i] {
                    valid_hours += 1;
                }
                if let Some(sr) = state_reset_mask {
                    if sr[i] {
                        has_reset = true;
                    }
                }
            }
        }
        if expected_hours == 0 {
            continue;
        }
        let missing_hours = expected_hours - valid_hours;
        let completeness = 100.0 * valid_hours as f64 / expected_hours as f64;

        let mut reasons = Vec::new();
        let is_incomplete_start = win_start < data_start && data_start <= win_end;
        let is_ongoing_latest = win_end > now;
        if is_incomplete_start {
            reasons.push("観測開始を含む不完全年".to_string());
        }
        if is_ongoing_latest {
            reasons.push("実行時点で終了していない最新年区分".to_string());
        }
        if completeness < completeness_threshold_percent {
            reasons.push(format!("データ完全率が{completeness_threshold_percent}%未満"));
        }
        if has_reset {
            reasons.push("大きな欠測により状態量が再初期化された区間を含む".to_string());
        }
        let is_eligible = reasons.is_empty();

        results.push(AnnualCompleteness {
            year_label: year_label(start_year, boundary),
            expected_hours,
            valid_hours,
            missing_hours,
            completeness_percent: completeness,
            has_state_reset: has_reset,
            is_eligible_default: is_eligible,
            exclusion_reasons: reasons,
        });
    }
    results
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn fiscal_year_window_boundaries() {
        let (start, end) = year_window(2025, &FISCAL_YEAR);
        use chrono::Datelike;
        assert_eq!((start.month(), start.day(), start.year()), (4, 1, 2025));
        assert_eq!((end.month(), end.day(), end.year()), (4, 1, 2026));
    }

    #[test]
    fn year_labels_are_formatted_correctly() {
        assert_eq!(year_label(2025, &CALENDAR_YEAR), "2025年");
        assert_eq!(year_label(2025, &FISCAL_YEAR), "2025年度");
        assert_eq!(year_label(2025, &JUNE_START_YEAR), "2025年6月始まり");
    }
}
